src/main/java/uk/gov/digital/ho/pttg/application/retry/NameMatchingCandidatesGenerator.java

```diff
@@ -0,0 +1,81 @@
+package uk.gov.digital.ho.pttg.application.retry;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class NameMatchingCandidatesGenerator {
+
+    // This maps the number of names to a list containing pairs of candidate names in the order defined by the requirements
+    private static final Map<Integer, List<List<Integer>>> CANDIDATE_NAME_RULES = ImmutableMap.of(
+            1, ImmutableList.of(
+                    ImmutableList.of(0, 0)),
+            2, ImmutableList.of(
+                    ImmutableList.of(0, 1), ImmutableList.of(1, 0)),
+            3, ImmutableList.of(
+                    ImmutableList.of(0, 2), ImmutableList.of(1, 2),
+                    ImmutableList.of(2, 0), ImmutableList.of(2, 1),
+                    ImmutableList.of(0, 1), ImmutableList.of(1, 0)),
+            4, ImmutableList.of(
+                    ImmutableList.of(0, 3), ImmutableList.of(1, 3), ImmutableList.of(2 ,3),
+                    ImmutableList.of(0, 1), ImmutableList.of(0, 2), ImmutableList.of(1, 0),
+                    ImmutableList.of(2, 0), ImmutableList.of(3, 0), ImmutableList.of(1, 2),
+                    ImmutableList.of(2, 1), ImmutableList.of(3, 1), ImmutableList.of(3, 2)),
+            5, ImmutableList.of(
+                    ImmutableList.of(0, 4), ImmutableList.of(1, 4), ImmutableList.of(2, 4), ImmutableList.of(3, 4),
+                    ImmutableList.of(0, 1), ImmutableList.of(0, 2), ImmutableList.of(0, 3), ImmutableList.of(1, 0),
+                    ImmutableList.of(1, 2), ImmutableList.of(1, 3), ImmutableList.of(2, 0), ImmutableList.of(2, 1),
+                    ImmutableList.of(2, 3), ImmutableList.of(3, 0), ImmutableList.of(3, 1), ImmutableList.of(3, 2),
+                    ImmutableList.of(4, 0), ImmutableList.of(4, 1), ImmutableList.of(4, 2), ImmutableList.of(4, 3))
+    );
+
+
+    public static List<String> generateCandidates(String firstName, String lastName) {
+        validateName(firstName, lastName);
+
+        List<String> candidates = new ArrayList<>();
+        List<String> allNames = findAllNames(firstName, lastName);
+
+        candidates.addAll(
+                CANDIDATE_NAME_RULES
+                        .get(allNames.size())
+                        .stream()
+                        .map(l -> allNames.get(l.get(0)) + " " + allNames.get(l.get(1)))
```

Comment:
I wonder if this could be clearer, `l` isn't very descriptive and there are a lot of indexes to keep track of which could become confusing.  Maybe it would be appropriate to create a CandidateNameRule/Pair class and use it something like this:
```java
.map(rule -> allNames.get(rule.firstIndex()) + " " + allNames.get(rule.secondIndex()));
```
or perhaps better and simpler is: 
```java
.map(pair -> allNames.get(pair.left()) + " " + allNames.get(pair.right())); 
```

Using Pair also means the big map above becomes clearer:

```java
ImmutableList.of(
    new Pair(0,2), new Pair(1,2),
)
```

Reply:
You could even turn it on it's head which might be more OO:

```java
.map(rule -> rule.apply(allNames))
```

I think that's the strategy pattern in fact, maybe it should be a CandidateNameStrategy?

```java
.map(strategy -> strategy.applyTo(allNames))
```

Reply:
Due to time constraints I've renamed l to something more representative.  Feel free to pick this up and make it nice if you like :)